riscv-executor/src/submachines.rs

```diff
@@ -44,152 +44,157 @@ pub trait Submachine<F: FieldElement> {
-    /// Consume the trace returning a list of columns.
-    /// Should be called only once.
-    fn take_cols(&mut self) -> Vec<(String, Vec<Elem<F>>)>;
+    fn add_operation(&mut self, selector_idx: Option<u8>, lookup_args: &[F], extra: &[F]);
+    /// finish the trace, padding to the given degree and returning the machine columns.
+    /// Ideally we'd take `self` here, but this is called from a `dyn Trait`...
+    fn finish(&mut self, degree: u32) -> Vec<(String, Vec<F>)>;
 }
 
 /// Concrete implementation of the Submachine trait
 struct SubmachineImpl<F: FieldElement, M: SubmachineKind> {
-    namespace: String,
     trace: SubmachineTrace<F>,
     m: std::marker::PhantomData<M>,
+    finished: bool,
 }
 
 impl<F: FieldElement, M: SubmachineKind> SubmachineImpl<F, M> {
     pub fn new(namespace: &str, witness_cols: &[String]) -> Self {
         // filter machine columns
         let prefix = format!("{namespace}::");
-        let witness_cols = witness_cols
+        let witness_cols: HashMap<_, _> = witness_cols
             .iter()
             .filter(|c| c.starts_with(namespace))
             .map(|c| (c.strip_prefix(&prefix).unwrap().to_string(), vec![]))
             .collect();
+        if witness_cols.is_empty() {
+            log::info!(
+                "namespace {} has no witness columns in the optimized pil",
+                namespace
+            );
+        }
         SubmachineImpl {
-            namespace: namespace.to_string(),
-            trace: SubmachineTrace::new(witness_cols),
+            trace: SubmachineTrace::new(namespace, witness_cols),
             m: std::marker::PhantomData,
+            finished: false,
         }
     }
 }
 
 impl<F: FieldElement, M: SubmachineKind> Submachine<F> for SubmachineImpl<F, M> {
     fn namespace(&self) -> &str {
-        self.namespace.as_str()
+        self.trace.namespace.as_str()
     }
 
     fn len(&self) -> u32 {
         self.trace.len()
     }
 
-    fn add_operation(
-        &mut self,
-        name: &str,
-        args: &[Elem<F>],
-        selector: Option<u32>,
-        submachines: &[&mut dyn Submachine<F>],
-    ) {
-        M::add_operation(&mut self.trace, name, args, selector, submachines);
-    }
-
-    fn final_row_override(&mut self) {
-        self.trace.final_row_override()
+    fn add_operation(&mut self, selector_idx: Option<u8>, lookup_args: &[F], extra: &[F]) {
+        M::add_operation(&mut self.trace, selector_idx, lookup_args, extra);
     }
 
-    fn push_dummy_block(&mut self, machine_max_degree: usize) {
-        let prev_len = self.len();
-        self.trace
-            .push_dummy_block(machine_max_degree, M::BLOCK_SIZE, M::SELECTORS);
-        let dummy_size = self.len() - prev_len;
-        M::dummy_block_fix(&mut self.trace, dummy_size);
-    }
-
-    fn take_cols(&mut self) -> Vec<(String, Vec<Elem<F>>)> {
+    fn finish(&mut self, degree: u32) -> Vec<(String, Vec<F>)> {
+        assert!(self.len() <= degree);
```

Comment:
we had a bug before related to this, where in the continuations dry run the executor is allowed to go to any amount of rows. Does this take that into account?